packages/aws-cdk/lib/api/garbage-collection/stack-refresh.ts

```diff
@@ -0,0 +1,180 @@
+import { CloudFormation } from 'aws-sdk';
+import { sleep } from '../../../test/util';
+import { debug } from '../../logging';
+
+export class ActiveAssetCache {
+  private readonly stacks: Set<string> = new Set();
+
+  public rememberStack(stackTemplate: string) {
+    this.stacks.add(stackTemplate);
+  }
+
+  public contains(asset: string): boolean {
+    for (const stack of this.stacks) {
+      if (stack.includes(asset)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+async function paginateSdkCall(cb: (nextToken?: string) => Promise<string | undefined>) {
+  let finished = false;
+  let nextToken: string | undefined;
+  while (!finished) {
+    nextToken = await cb(nextToken);
+    if (nextToken === undefined) {
+      finished = true;
+    }
+  }
+}
+
+/** We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+ * If we encounter this status, we will wait up to the maxWaitTime before erroring out
+ */
+async function listStacksNotBeingReviewed(cfn: CloudFormation, maxWaitTime: number, nextToken: string | undefined) {
+  let sleepMs = 500;
+
+  while (sleepMs < maxWaitTime) {
+    let stacks = await cfn.listStacks({ NextToken: nextToken }).promise();
+    if (!stacks.StackSummaries?.some(s => s.StackStatus == 'REVIEW_IN_PROGRESS')) {
+      return stacks;
+    }
+    await sleep(Math.floor(Math.random() * sleepMs));
+    sleepMs = Math.min(sleepMs * 2, maxWaitTime);
+  }
+
+  throw new Error(`Stacks still in REVIEW_IN_PROGRESS state after waiting for ${maxWaitTime} ms.`);
+}
+
+/**
+   * Fetches all relevant stack templates from CloudFormation. It ignores the following stacks:
+   * - stacks in DELETE_COMPLETE or DELETE_IN_PROGRES stage
+   * - stacks that are using a different bootstrap qualifier
+   *
+   * It fails on the following stacks because we cannot get the template and therefore have an imcomplete
+   * understanding of what assets are being used.
+   * - stacks in REVIEW_IN_PROGRESS stage
+   */
+async function fetchAllStackTemplates(cfn: CloudFormation, maxWaitTime: number, qualifier?: string) {
+  const stackNames: string[] = [];
+  await paginateSdkCall(async (nextToken) => {
+    const stacks = await listStacksNotBeingReviewed(cfn, maxWaitTime, nextToken);
+
+    // Deleted stacks are ignored
+    const ignoredStatues = ['CREATE_FAILED', 'DELETE_COMPLETE', 'DELETE_IN_PROGRESS', 'DELETE_FAILED'];
```

Comment:
🗿

Reply:
gonna assume you aren't a fan of my comment so i updated it